### PR TITLE
DAOS-16621 build: Fix Go versions in rpm/deb packaging

### PR DESCRIPTION
**Pull request by kjacque**
- Set Go minimum version to 1.21 in rpm and debian packaging spec files.
- Update scons Go version check to use version in go.mod.
- Add a reminder in go.mod file so we remember the packaging files when bumping the minimum Go version in the future.
- Update Ubuntu 22.04 Dockerfile to get an appropriate version of Go.

Required-githooks: true

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by kjacque** [comment]
No need to run feature tests -- these changes affect the build only.

File: debian/control
Comment (ashleypittman) [inline debian/control:28]: This has come up before and got dropped but I think the golang-go version on debian uses an epoch so you need to specify the version as I've indicated here.
```suggestion
               golang-go (>= 2:1.21),
```

File: site_scons/site_tools/go_builder.py
Comment (ashleypittman) [inline site_scons/site_tools/go_builder.py:1]: There is a lot more code that can be removed from this file if you're removing the go version check.  As this PR stands it'll load the go version but then not do anything with it.

**Review by ashleypittman — CHANGES_REQUESTED** [review]
I don't know if you're aware but docker builds on Ubuntu are failing, as per the build log at https://github.com/daos-stack/daos/actions/runs/10962467337/job/30441777476

What is happening there is the default go version is 1.18, the scons check is detecting this and passing, then there's a non-obvious build failure later on.  In light of that perhaps we need to keep the scons check for now (but update the target version obviously) but also we need to update the packages for that build somehow.

File: debian/control
Comment (ashleypittman) [reply debian/control:28]: https://packages.debian.org/bookworm/golang-go for the upstream package, and yes it does have/require an epoch number.

File: site_scons/site_tools/go_builder.py
Comment (kjacque) [reply site_scons/site_tools/go_builder.py:1]: Given the weirdness in the failed build you pointed out, maybe I'll leave the scons check in. The reason I left some of that code in was because it looked like it was printing it to the console, which is useful for debugging build issues.

**Comment by kjacque** [comment]
> In light of that perhaps we need to keep the scons check for now (but update the target version obviously) but also we need to update the packages for that build somehow.

Ubuntu LTS versions tend to stick at a very old version of Go, to the point where I think it's more common to either download/install Go manually or use a third-party PPA. The longsleep/go-backports PPA is the one I've seen used most often, and is updated on a regular cadence. Do we need to get approval for that? @brianjmurrell what do you think?

**Comment by mjmac** [comment]
> Ubuntu LTS versions tend to stick at a very old version of Go, to the point where I think it's more common to either download/install Go manually or use a third-party PPA. The longsleep/go-backports PPA is the one I've seen used most often, and is updated on a regular cadence. Do we need to get approval for that? @brianjmurrell what do you think?

A better option than the PPA is to just install the official golang-go packages from the next LTS release. There is some discussion about the best way to do this in another PR: https://github.com/daos-stack/daos/pull/15150/files#diff-44f4cb066190d53259ef4c6c84dbac0f295f7e229ddb213000deadb5b9f7cb9eR17

File: utils/rpms/daos.spec
Comment (mjmac) [inline utils/rpms/daos.spec:595]: Looks like the version got goofed up here.

File: site_scons/site_tools/go_builder.py
Comment (mjmac) [inline site_scons/site_tools/go_builder.py:11]: May be overengineering, but in the interest of minimizing the number of places that need to be updated, maybe this could be based on the version in `go.mod`?

Incidentally, for the record, I do agree that we should keep this scons check for the clear failure when building from source with an outdated Go toolchain.

File: utils/rpms/daos.spec
Comment (kjacque) [reply utils/rpms/daos.spec]: Good catch. Fixed.

File: site_scons/site_tools/go_builder.py
Comment (kjacque) [reply site_scons/site_tools/go_builder.py:11]: I did this and it seems to work. Thanks, this should make things a little easier to maintain in the future.

File: site_scons/site_tools/go_builder.py
Comment (kjacque) [reply site_scons/site_tools/go_builder.py:1]: I ended up taking Mike's suggestion below and automating the minimum version detection based on go.mod.

**Comment by kjacque** [comment]
> Ubuntu LTS versions tend to stick at a very old version of Go, to the point where I think it's more common to either download/install Go manually or use a third-party PPA. The longsleep/go-backports PPA is the one I've seen used most often, and is updated on a regular cadence. Do we need to get approval for that? @brianjmurrell what do you think?

Latest commits use the external PPA I mentioned here.

File: site_scons/site_tools/go_builder.py
Comment (mjmac) [inline site_scons/site_tools/go_builder.py:58]: Hmm. I don't think that it's safe to unconditionally append the `.0` suffix here. Per [the spec](https://go.dev/ref/mod#go-mod-file-go), the `go` directive may contain any valid [Go version](https://go.dev/doc/toolchain#version).

After looking at little more closely at the version comparison, I remembered that it doesn't rely on the versions being balanced, i.e. it works fine if one is `x.y.z` and one is `x.y`, so you could just return `parts[1]` here and be done with it, except for the fact that the version might contain an `rc` string. Sigh. You've got the `get_go_version()` helper below that almost does what you need, except that it won't quite match on either the line or just the version. Were it me, I'd probably just do something like `return get_go_version("go" + parts[1])` and call it a day.

The other thing is that if this function somehow fails to get a version from that file, returning `None` will make the version comparison fail with a mysterious error like `AttributeError: 'NoneType' object has no attribute 'split'`. Might be best to just raise a specific exception instead.

Comment (kjacque) [reply site_scons/site_tools/go_builder.py]: Good idea. Thanks. I don't know how likely it is for us to use more complicated go version strings in the go.mod file, but simple enough to harden it up.

File: site_scons/site_tools/go_builder.py
Comment (ashleypittman) [inline site_scons/site_tools/go_builder.py:59]: @jolivier23 might remember quicker than me but there's a `scons` exception we cal call here I think.  Either way the error needs to be resolved before this can be landed.

**Review by ashleypittman — COMMENTED** [review]
Without having Ubuntu in the support matrix (which I don't think it is) then Ubuntu dockerfiles are for convenience only and I see no reason why we shouldn't be updating to the latest stable or LTS release.  This dockerfile is only invoked from GHA that I'm aware of.

File: site_scons/site_tools/go_builder.py
Comment (ashleypittman) [reply site_scons/site_tools/go_builder.py:59]: Reading this through again, there's a scons configure check going on here but the configure check itself is also only called from this file and if the configure check fails then the build stops with `Exit(1)` (a scons function).

What I think you need to do here is return `None` here, and then when calling this function later on check for `None` and call `context.Result` with the error string and return 0.

File: site_scons/site_tools/go_builder.py
Comment (ashleypittman) [inline site_scons/site_tools/go_builder.py:95]:
```suggestion
        min_go_version = get_min_go_version()
        if min_go_version is None:
            context.Result("no minimum go version found in go.mod")
            return 1
```
This will need testing and adjusting for whitespace I assume.

File: site_scons/site_tools/go_builder.py
Comment (ashleypittman) [inline site_scons/site_tools/go_builder.py:94]: You're not checking the minimum version of the installed binary, you're checking the binary against a minimum version of the language.
```suggestion
        context.Display('Getting minimum go version... ')
```
As this code stands it's outputting the below which feels incorrect.
```
Checking for Go compiler... /bin/go
Getting minimum /bin/go version... 1.21
Checking /bin/go version... 1.21.13
```

File: site_scons/site_tools/go_builder.py
Comment (kjacque) [reply site_scons/site_tools/go_builder.py:95]: Yeah, this was similar to what I had before. I'll change it back.

File: site_scons/site_tools/go_builder.py
Comment (kjacque) [reply site_scons/site_tools/go_builder.py:94]: Fair point, fixing.

**Comment by ashleypittman** [comment]
The other thing on the scons check front is that it might be possible to just get the `go` command to do "something" with the go.mod file and see if fails or not.  I'm not sure if there is a suitable go command for this that exists in 1.21 though.  It would simplify all the code we have if we could just do a test compile.  Out of scope for this PR though.

File: site_scons/site_tools/go_builder.py
Comment (ashleypittman) [reply site_scons/site_tools/go_builder.py:95]: I missed that, sorry, but I think this is the correct way to fail the build if the check cannot be run.

File: utils/scripts/install-ubuntu.sh
Comment (ashleypittman) [inline utils/scripts/install-ubuntu.sh:20]: I thought I'd made a comment on this change but I can't see it anywhere, apologies if I'm repeating myself here.

One of the intentions of these scripts is that an admin should be able to call them to install required packages and to do so safely, so therefore they install packages without the -y option for example.  The package manager/repo configuration is all done in the dockerfile/helper scripts and this script then just installs the named packages.

This code breaks that model so the script will re-configure the node as well as install code and in addition it's also possibly not idempotent any more.

Comment (kjacque) [reply utils/scripts/install-ubuntu.sh:20]: I tested moving the Dockerfile Ubuntu version forward to 24.04 LTS and ran into a warning with the interception library (src/client/dfuse/il), I think due to an upgraded version of gcc.

```
In file included from src/client/dfuse/il/int_posix.c:32:
src/client/dfuse/il/intercept.h:38:21: error: ‘fseek64’ specifies less restrictive attribute than its target ‘fseek’: ‘nonnull’ [-Werror=missing-attributes]
   38 |         ACTION(int, fseek, (FILE *, long, int))                                                    \
      |                     ^~~~~
src/client/dfuse/il/intercept.h:106:27: note: in definition of macro ‘IOIL_DECLARE_ALIAS64’
  106 |         DFUSE_PUBLIC type name##64 params __attribute__((weak, alias(#name)));
      |                           ^~~~
src/client/dfuse/il/int_posix.c:3023:1: note: in expansion of macro ‘FOREACH_ALIASED_INTERCEPT’
 3023 | FOREACH_ALIASED_INTERCEPT(IOIL_DECLARE_ALIAS64)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~
In file included from src/client/dfuse/il/int_posix.c:21:
/usr/include/stdio.h:779:12: note: ‘fseek64’ target declared here
  779 | extern int fseek (FILE *__stream, long int __off, int __whence)
      |            ^~~~~
cc1: all warnings being treated as errors
```

I think pushing up the Ubuntu version is out of scope for this task, given that it'll require production code changes.

I could shift the PPA addition to the Dockerfile if that would be an acceptable compromise.

**Review by mjmac — APPROVED** [review]

**Review by tanabarr — APPROVED** [review]

**Review by mjmac — APPROVED** [review]

File: utils/scripts/install-ubuntu.sh
Comment (ashleypittman) [reply utils/scripts/install-ubuntu.sh]: If there's build errors then we definitely should update, that's what we're looking for with this test.  Outside the scope of this PR though but I'll file a ticket and take it on.

File: src/control/go.mod
Comment (ashleypittman) [inline src/control/go.mod:6]: I'm not familiar with go, what does this part of the change do?

**Review by ashleypittman — APPROVED** [review]

File: src/control/go.mod
Comment (kjacque) [reply src/control/go.mod:6]: Yeah, the go mod tools added that line automatically, but I don't think we need it. It seems that it's only relevant if we want to use a different toolchain version than the minimum go version.

**Pull request merged**
